src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -5,51 +5,202 @@
 import com.sun.net.httpserver.HttpHandler;
 import com.sun.net.httpserver.HttpServer;
+import com.sun.net.httpserver.spi.HttpServerProvider;
 
 import java.io.IOException;
+import java.net.BindException;
 import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketException;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
 import java.util.concurrent.ExecutorService;
-import javax.inject.Inject;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.net.SocketFactory;
 
 public class HttpServerImpl implements Server {
-    public static final String URL = "/2016-11-01/credentialprovider/";
+
+    private static final Logger logger = LogManager.getLogger(HttpServerImpl.class);
+    static final String URL = "/2016-11-01/credentialprovider/";
     private static final int TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS = 1;
-    private final HttpServer httpImpl;
+    @SuppressWarnings("PMD.AvoidUsingHardCodedIP")
+    private static final String ADDR_IPV4 = "127.0.0.1";
+    @SuppressWarnings("PMD.AvoidUsingHardCodedIP")
+    private static final String ADDR_IPV6 = "::1";
 
-    @Inject
-    private final HttpHandler credentialRequestHandler;
+    /**
+     * HTTP server port, provided by configuration.
+     *
+     * <p>The actual HTTP server port may be different from this,
+     * if configured port is system-assigned (0).
+     */
+    private final int configuredPort;
 
-    @Inject
+    private HttpServer serverIPv4;
+    private HttpServer serverIPv6;
+
+    private final HttpHandler credentialRequestHandler;
     private final ExecutorService executorService;
+    private final SocketFactory socketFactory;
+    private final HttpServerProvider httpServerProvider;
 
     /**
      * Constructor.
-     * @param port Http server port
+     * @param port HTTP server port
      * @param credentialRequestHandler request handler for server requests
      * @param executorService executor service instance
      * @throws IOException When server creation fails
      */
     HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) throws IOException {
-        httpImpl = HttpServer.create(new InetSocketAddress("localhost", port), 0);
+        this(port, credentialRequestHandler, executorService,
+                HttpServerProvider.provider(), SocketFactory.getDefault());
+    }
+
+    HttpServerImpl(int port, HttpHandler credentialRequestHandler,
+                   ExecutorService executorService, HttpServerProvider httpServerProvider, SocketFactory socketFactory)
+            throws IOException {
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.httpServerProvider = httpServerProvider;
+        this.socketFactory = socketFactory;
+        createServers(configuredPort);
```

Comment:
having this is the constructor scares me. Creating and starting the server should be moved into `start`.

Reply:
I did at one point 😅 
will revise and make sure start is idempotent